webapi/src/main/scala/org/knora/webapi/responders/v1/UsersResponderV1.scala

```diff
@@ -35,123 +41,431 @@ import scala.concurrent.Future
+        for {
+            sparqlQueryString <- Future(queries.sparql.v1.txt.getUserByEmail(
+                triplestore = settings.triplestoreType,
+                email = email
+            ).toString())
+            //_ = log.debug(s"getUserProfileByEmailV1 - sparqlQueryString: $sparqlQueryString")
+
+            userDataQueryResponse <- (storeManager ? SparqlSelectRequest(sparqlQueryString)).mapTo[SparqlSelectResponse]
+
+            //_ = log.debug(MessageUtil.toSource(userDataQueryResponse))
 
-            projectIris = groupedUserData.get(OntologyConstants.KnoraBase.IsInProject) match {
-                case Some(projects) => projects
-                case None => Vector.empty[IRI]
+            _ = if (userDataQueryResponse.results.bindings.isEmpty) {
+                throw NotFoundException(s"User '$email' not found")
             }
 
-        } yield UserProfileV1(userDataV1, groupIris, projectIris)
+            userProfileV1 <- userDataQueryResponse2UserProfile(userDataQueryResponse, profileType)
+
+        } yield userProfileV1 // UserProfileV1(userDataV1, groupIris, projectIris)
     }
 
     /**
-      * Gets information about a Knora user, and returns it in a [[UserProfileV1]].
+      * Creates a new user. Self-registration is allowed, so even the default user, i.e. with no credentials supplied,
+      * is allowed to create a new user.
       *
-      * @param username the username of the user.
-      * @return a [[UserProfileV1]] describing the user.
+      * Referenced Websites:
+      *                     - https://crackstation.net/hashing-security.htm
+      *                     - http://blog.ircmaxell.com/2012/12/seven-ways-to-screw-up-bcrypt.html
+      *
+      * @param createRequest a [[CreateUserApiRequestV1]] object containing information about the new user to be created.
+      * @param userProfile   a [[UserProfileV1]] object containing information about the requesting user.
+      * @return a future containing the [[UserOperationResponseV1]].
       */
-    private def getUserProfileByUsernameV1(username: String): Future[UserProfileV1] = {
+    private def createNewUserV1(createRequest: CreateUserApiRequestV1, userProfile: UserProfileV1, apiRequestID: UUID): Future[UserOperationResponseV1] = {
+
         for {
-            sparqlQuery <- Future(queries.sparql.v1.txt.getUserByUsername(
+            // check if required information is supplied
+            _ <- Future(if (createRequest.email.isEmpty) throw BadRequestException("Email cannot be empty"))
+            _ = if (createRequest.password.isEmpty) throw BadRequestException("Password cannot be empty")
+            _ = if (createRequest.givenName.isEmpty) throw BadRequestException("Given name cannot be empty")
+            _ = if (createRequest.familyName.isEmpty) throw BadRequestException("Family name cannot be empty")
+
+
+            // check if the supplied email for the new user is unique, i.e. not already registered
```

Comment:
This looks like it has a race condition: if two requests try to create the same user at the same time, this check will pass for both of them. Or if one request is changing an existing user's email address from `foo@bar.org` to `foo@example.org`, and another request simultaneously tries to create a new user `foo@example.org`. One solution would be to have all update operations on users lock a single IRI, like `http://data.knora.org/users`. This could be a bottleneck, but updates on users are probably fast operations, so maybe it would be OK.